### PR TITLE
Add http stats collection for generic http client.

### PR DESCRIPTION
**Pull request by rghetia**

### CODE REVIEW

**Comment by rghetia** [comment]
@bogdandrutu PTAL

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 1]
```diff
@@ -47,7 +50,7 @@
    * @param type the {@code MessageEvent.Type} of the event.
    * @param uncompressedMessageSize size of the message before compressed (optional).
    * @param compressedMessageSize size of the message after compressed (optional).
-   * @since 0.18
+   * @since 0.19
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Please send a separate PR to change all "since" tags that are not released.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 2]
```diff
@@ -66,36 +69,33 @@ private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullabl
...
-  // [TODO:rghetia] add it back after 0.18 is released
-  final void handleMessageSent(Span span, long messageId, long messageSize) {
-    checkNotNull(span, "span");
-    if (span.getOptions().contains(Options.RECORD_EVENTS)) {
+  public final void handleMessageSent(HttpContext context) {
+    checkNotNull(context, "context");
+    if (context.span.getOptions().contains(Options.RECORD_EVENTS)) {
       // record compressed size
-      recordMessageEvent(span, messageId, Type.SENT, messageSize, 0L);
+      recordMessageEvent(context.span, context.reqId, Type.SENT, context.sentMessageSize.get(), 0L);
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Seems to me that reqId is not what should be used here. The message id should be a unique id (for each message sent/received in a stream). reqId seems to be constant for the entire request.

So what we need to record is the following:
message sent from client {sent, id = 1, size} - on server {recv, id = 1, size}
message recv by client {recv, id = 1, size} - on server {send, id = 1, size}
message recv by client {recv, id = 2, size} - on server {send, id = 2, size}
message recv by client {recv, id = 3, size} - on server {send, id = 3, size}

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpContext.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import io.opencensus.trace.Span;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class provides storage per request context on http client and server.
+ *
+ * @since 0.19
+ */
+@ExperimentalApi
+public class HttpContext {
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpContext.java]: Maybe HttpRequestContext?

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java [hunk 4]
```diff
@@ -42,8 +42,8 @@
  * @since 0.18
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java]: Forgot to modify since. Better to do this in a separate PR and remove all todos like this.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 1 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: ok

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 2 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: so the even is not for when the entire request/response content is sent/received but rather the individual chunks. 
What is the significance of the id? How is it used?

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java (hunk 4 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpServerHandler.java]: ok.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpContext.java (hunk 3 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpContext.java]: ok.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 2 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: I have changed it to record the event per message sent/received instead of per request.

**Comment by rghetia** [comment]
@bogdandrutu PTAL.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 5]
```diff
@@ -66,36 +69,34 @@ private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullabl
...
   /**
-   * Instrument an HTTP span after a message is sent.
+   * Instrument an HTTP span after a message is sent. Typically called when last chunk of request or
+   * response is sent.
    *
-   * @param span the span.
-   * @param messageId an id for the message.
-   * @param messageSize the size of the message.
+   * @param context request specific {@link HttpRequestContext}
    * @since 0.18
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Please send me the PR to fix all since labels as I commented before.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java [hunk 6]
```diff
@@ -66,36 +69,34 @@ private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullabl
...
-  final void handleMessageReceived(Span span, long messageId, long messageSize) {
-    checkNotNull(span, "span");
-    if (span.getOptions().contains(Options.RECORD_EVENTS)) {
+  public final void handleMessageReceived(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context");
+    context.receiveMessageSize.addAndGet(bytes);
+    if (context.span.getOptions().contains(Options.RECORD_EVENTS)) {
       // record compressed size
-      recordMessageEvent(span, messageId, Type.RECEIVED, messageSize, 0L);
+      recordMessageEvent(context.span, context.reqId, Type.RECEIVED, bytes, 0L);
```
Comment (bogdandrutu) [inline contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: Should we have to ids one for received and one for sent messages?

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 5 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: done.
https://github.com/census-instrumentation/opencensus-java/pull/1636

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java (hunk 6 above)
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: what is the purpose of ids? Is it more like a sequence number rather than an id? Not sure how it will be used

Comment (bogdandrutu) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: You need to match the message sent with the message received in order to determine network issues or other things.

Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/AbstractHttpHandler.java]: done.

File: contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java [hunk 7]
```diff
@@ -107,20 +124,56 @@ public Span handleStart(@Nullable Span parent, C carrier, Q request) {
...
+      HttpRequestContext context,
+      @Nullable Q request,
+      @Nullable P response,
+      @Nullable Throwable error) {
+    double requestLatency = NANOSECONDS.toMillis(System.nanoTime() - context.requestStartTime);
+
+    String methodStr = request == null ? "" : extractor.getMethod(request);
+    TagContext startCtx =
+        tagger
+            .currentBuilder()
```
[the remark this answers is not in the document]
Comment (rghetia) [reply contrib/http_util/src/main/java/io/opencensus/contrib/http/HttpClientHandler.java]: sure.
